From 72be4bb967e8081d4864619ff2afaad77e1956e2 Mon Sep 17 00:00:00 2001
From: Vikram Kandukuri <quic_vikram@quicinc.com>
Date: Wed, 20 Oct 2021 23:10:53 +0530
Subject: [PATCH] iw: Add MLO support in interface creation

Add MLO support in interface creation and add support to get
MLD mac address in dev info.

Signed-off-by: Vikram Kandukuri <quic_vikram@quicinc.com>
---
 interface.c | 27 ++++++++++++++++++++++++++-
 nl80211.h   |  3 +++
 2 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/interface.c b/interface.c
index 57e2fe4..88b30e9 100644
--- a/interface.c
+++ b/interface.c
@@ -236,7 +236,9 @@ static int handle_interface_add(struct nl80211_state *state,
 	enum nl80211_iftype type;
 	int tpset;
 	unsigned char mac_addr[ETH_ALEN];
-	int found_mac = 0;
+	unsigned char mld_mac_addr[ETH_ALEN];
+	int found_mac = 0, found_mld_mac = 0;
+	char mac_temp[20];
 
 	if (argc < 1)
 		return 1;
@@ -287,6 +289,18 @@ try_another:
 				fprintf(stderr, "flags error\n");
 				return 2;
 			}
+		} else if (strcmp(argv[0], "mld_addr") == 0) {
+			argc--;
+			argv++;
+			if (mac_addr_a2n(mld_mac_addr, argv[0])) {
+				fprintf(stderr, "Invalid MAC address\n");
+				return 2;
+			}
+			argc--;
+			argv++;
+			found_mld_mac = 1;
+			mac_addr_n2a(mac_temp, mld_mac_addr);
+			goto try_another;
 		} else {
 			return 1;
 		}
@@ -302,6 +316,10 @@ try_another:
 	if (found_mac)
 		NLA_PUT(msg, NL80211_ATTR_MAC, ETH_ALEN, mac_addr);
 
+	if (found_mld_mac)
+		NLA_PUT(msg, NL80211_ATTR_MLD_MAC, ETH_ALEN, mld_mac_addr);
+
+
 	return 0;
  nla_put_failure:
 	return -ENOBUFS;
@@ -399,6 +417,13 @@ static int print_iface_handler(struct nl_msg *msg, void *arg)
 		mac_addr_n2a(mac_addr, nla_data(tb_msg[NL80211_ATTR_MAC]));
 		printf("%s\taddr %s\n", indent, mac_addr);
 	}
+
+	if (tb_msg[NL80211_ATTR_MLD_MAC]) {
+		char mld_mac_addr[20];
+		mac_addr_n2a(mld_mac_addr, nla_data(tb_msg[NL80211_ATTR_MLD_MAC]));
+		printf("%s\tmld_addr %s\n", indent, mld_mac_addr);
+	}
+
 	if (tb_msg[NL80211_ATTR_SSID]) {
 		printf("%s\tssid ", indent);
 		print_ssid_escaped(nla_len(tb_msg[NL80211_ATTR_SSID]),
diff --git a/nl80211.h b/nl80211.h
index 174851b..5654aca 100644
--- a/nl80211.h
+++ b/nl80211.h
@@ -3047,6 +3047,9 @@ enum nl80211_attrs {
 
 	NL80211_ATTR_WIPHY_ANTENNA_GAIN,
 
+	NL80211_ATTR_MLD_MAC=352,
+	NL80211_ATTR_MLD_REFERENCE=353,
+
 	/* add attributes here, update the policy in nl80211.c */
 
 	__NL80211_ATTR_AFTER_LAST,
-- 
2.7.4

